libafl_qemu/src/injections.rs

```diff
@@ -0,0 +1,312 @@
+    _state: Option<&mut S>,
+    syscall: i32, // syscall number
+    x0: u64,      // registers ...
+    x1: u64,
+    _x2: u64,
+    _x3: u64,
+    _x4: u64,
+    _x5: u64,
+    _x6: u64,
+    _x7: u64,
+) -> SyscallHookResult
+where
+    QT: QemuHelperTuple<S>,
+    S: UsesInput,
+{
+    //println!("syscall_hook {} {}", syscall, SYS_execve);
+    debug_assert!(i32::try_from(SYS_execve).is_ok());
+    if syscall == SYS_execve as i32 {
+        let _helper = hooks
+            .helpers_mut()
+            .match_first_type_mut::<QemuInjectionHelper>()
+            .unwrap();
+        if x0 > 0 && x1 > 0 {
+            let c_array = x1 as *const *const c_char;
+            let cmd = unsafe {
+                let c_str_ptr = x0 as *const c_char;
+                CStr::from_ptr(c_str_ptr).to_string_lossy()
+            };
+            assert!(
+                cmd.to_lowercase() != "fuzz",
+                "Found verified command injection!"
+            );
+            //println!("CMD {}", cmd);
+
+            let first_parameter = unsafe {
+                if (*c_array.offset(1)).is_null() {
+                    return SyscallHookResult::new(None);
+                }
+                CStr::from_ptr(*c_array.offset(1)).to_string_lossy()
+            };
+            let second_parameter = unsafe {
+                if (*c_array.offset(2)).is_null() {
+                    return SyscallHookResult::new(None);
+                }
+                CStr::from_ptr(*c_array.offset(2)).to_string_lossy()
+            };
+            if first_parameter == "-c"
+                && (second_parameter.to_lowercase().contains("';fuzz;'")
+                    || second_parameter.to_lowercase().contains("\";fuzz;\""))
+            {
+                panic!("Found command injection!");
+            }
+
+            //println!("PARAMETERS First {} Second {}", first_parameter, second_
+        }
+        SyscallHookResult::new(Some(0))
+    } else {
+        SyscallHookResult::new(None)
+    }
+}
+
+fn find_function(
+    emu: &Emulator,
+    file: &String,
+    function: &str,
+    loadaddr: GuestAddr,
+) -> Result<GuestAddr, Error> {
+    let mut elf_buffer = Vec::new();
+    let elf = EasyElf::from_file(file, &mut elf_buffer)?;
+    let offset = if loadaddr > 0 {
+        loadaddr
+    } else {
+        emu.load_addr()
+    };
+    let start_pc = elf
+        .resolve_symbol(function, offset)
+        .ok_or_else(|| Error::empty_optional("Symbol {function} not found in {file}"))?;
+    println!("Found {function} in {file}");
+    Ok(start_pc)
+}
+
+extern "C" fn on_call_check(val: u64, _pc: GuestAddr) {
+    let emu = Emulator::new_empty();
+    let parameter: u8 = (val & 0xff) as u8;
+    let off: usize = (val >> 8) as usize;
+
+    //println!("on_call_check {} {}", parameter, off);
+
+    #[cfg(any(target_arch = "x86_64", not(target_arch)))]
+    let reg: GuestAddr = match parameter {
+        0 => emu.current_cpu().unwrap().read_reg(Regs::Rdi).unwrap_or(0),
```

Comment:
Do we really want to default to `0` and not somehow error in this case?

Reply:
Also, you could write this more concise with something like
```rust
let param_id = match parameter { ... };
let reg = emu.current_cpu().unwrap().read_reg(param_id);
```

Reply:
Yes, 0 is fine. it is the same as null pointer, so it makes sense to use it.
When I try the param_id = ... approach then it does not compile anymore, whatever I try.